UM/View/GL/ShaderProgram.py

```diff
@@ -0,0 +1,180 @@
+
+import configparser
+import ast
+
+##  Raised when an error occurs during loading of the shader file.
+class InvalidShaderProgramError(Exception):
+    pass
+
+##  An abstract class for dealing with shader programs.
+#
+#   This class provides an interface an some basic elements for dealing with
+#   shader programs. Shader programs are described in a simple text file
+#   based on the Python configparser module. These files contain the shaders
+#   for the different shader program stages, in addition to defaults that should
+#   be used for uniform values and uniform and attribute bindings.
+class ShaderProgram:
+    def __init__(self):
+        self._bindings = {}
+        self._attribute_bindings = {}
+
+    ##  Load a shader program file.
+    #
+    #   This method loads shaders from a simple text file, using Python's configparser
+    #   as parser.
+    #
+    #   \note When writing shader program files, please note that configparser expects
+    #   indented lines for multiline values. Since the shaders are provided as a single
+    #   multiline string, make sure to indent them properly.
+    #
+    #   \param file_name The shader file to load.
+    #
+    #   \exception{InvalidShaderProgramError} Raised when the file provided does not contain any shaders.
+    def load(self, file_name):
+        parser = configparser.ConfigParser(interpolation = None)
+        parser.optionxform = lambda option: option
+        parser.read(file_name)
+
+        if not "shaders" in parser:
+            raise InvalidShaderProgramError("{0} is missing a vertex of fragment shader".format(file_name))
+
+        if not "vertex" in parser["shaders"] or not "fragment" in parser["shaders"]:
+            raise InvalidShaderProgramError("{0} is missing a vertex of fragment shader".format(file_name))
+
+        self.setVertexShader(parser["shaders"]["vertex"])
+        self.setFragmentShader(parser["shaders"]["fragment"])
+
+        self.build()
+
+        if "defaults" in parser:
+            for key, value in parser["defaults"].items():
+                self.setUniformValue(key, ast.literal_eval(value), cache = True)
+
+        if "bindings" in parser:
+            for key, value in parser["bindings"].items():
+                self.addBinding(key, value)
+
+        if "attributes" in parser:
+            for key, value in parser["attributes"].items():
+                self.addAttributeBinding(key, value)
+
+    ##  Set the vertex shader to use.
+    #
+    #   \param shader \type{string} The vertex shader to use.
+    def setVertexShader(self, shader):
+        raise NotImplementedError()
+
+    ##  Set the fragment shader to use.
+    #
+    #   \param shader \type{string} The fragment shader to use.
+    def setFragmentShader(self, shader):
+        raise NotImplementedError()
+
+    ##  Build the complete shader program out of the separately provided sources.
+    def build(self):
+        raise NotImplementedError()
```

Comment:
Add the same warning as you did for other files ("Should be implemented by subclasses")